usr/bs_hyc.c

```diff
@@ -0,0 +1,388 @@
+			assert(resultsp[i].result == 0);
+			target_cmd_io_done(hcmdp->cmdp, SAM_STAT_GOOD);
+
+			DLL_REM(&hcmdp->list);
+			free(hcmdp);
+		}
+		pthread_mutex_unlock(&infop->lock);
+		memset(resultsp, 0, sizeof (*resultsp) * infop->nr_results);
+	}
+}
+
+static int bs_hyc_open(struct scsi_lu *lup, char *pathp,
+			int *fdp, uint64_t *sizep)
+{
+	struct bs_hyc_info *infop = BS_HYC_I(lup);
+	int                 rc = 0;
+	int                 ffd = -1;
+	int                 efd = -1;
+	uint32_t            blksize;
+
+	ffd = backed_file_open(pathp, O_RDWR | O_LARGEFILE | O_DIRECT, sizep,
+		&blksize);
+	if (ffd < 0) {
+		eprintf("Failed to open %s, for tgt: %d, lun: %"PRId64 ", %m\n",
+			pathp, infop->lup->tgt->tid, infop->lup->lun);
+		rc = ffd;
+		goto error;
+	}
+
+	/** TODO: Validate blksize, whether its what we need or not */
+	if (!lup->attrs.no_auto_lbppbe) {
+		update_lbppbe(lup, blksize);
+	}
+
+	efd = eventfd(0, O_NONBLOCK);
+	if (efd < 0) {
+		rc = errno;
+		goto error;
+	}
+
+	rc = tgt_event_add(efd, EPOLLIN, bs_hyc_handle_completion, infop);
+	if (rc < 0) {
+		goto error;
+	}
+
+	infop->done_eventfd = efd;
+
+	assert(infop->vmdk != kInvalidVmdkHandle);
+	rc = SetVmdkEventFd(infop->vmdk, infop->done_eventfd);
+	if (rc < 0) {
+		goto error;
+	}
+
+	return 0;
+error:
+	if (efd >= 0) {
+		close(efd);
+		efd = -1;
+	}
+	if (ffd >= 0) {
+		close(ffd);
+		ffd = -1;
+	}
+	return rc;
+}
+
+static void bs_hyc_close(struct scsi_lu *lup)
+{
+	struct bs_hyc_info *infop = BS_HYC_I(lup);
+
+	assert(infop);
+	assert(infop->done_eventfd >= 0);
+
+	tgt_event_del(infop->done_eventfd);
+	close(infop->done_eventfd);
+	infop->done_eventfd = -1;
+
+	close(lup->fd);
+}
+
+enum {
+	Opt_vmdkid, Opt_err,
+};
+
+static match_table_t bs_hyc_opts = {
+	{Opt_vmdkid, "vmdkid=%s"},
+	{Opt_err, NULL},
+};
+
+static tgtadm_err bs_hyc_init(struct scsi_lu *lup, char *bsoptsp)
+{
+	struct bs_hyc_info *infop = BS_HYC_I(lup);
+	tgtadm_err          e = TGTADM_SUCCESS;
+	int                 rc;
+	char               *p;
+	char               *vmdkid = NULL;
+
+	while((p = strsep(&bsoptsp, ",")) != NULL) {
+		substring_t args[MAX_OPT_ARGS];
+		int token;
+		if (!*p)
+			continue;
+		token = match_token(p, bs_hyc_opts, args);
+		switch (token) {
+		case Opt_vmdkid:
+			vmdkid = match_strdup(&args[0]);
+		default:
+			break;
+		}
+	}
+	if (!vmdkid) {
+		eprintf("hyc bst needs vmdkid as bsopts\n");
+		return TGTADM_INVALID_REQUEST;
+	}
+
+	memset(infop, 0, sizeof(*infop));
+
+	infop->lup = lup;
+
+	assert(lup->tgt);
+	assert(lup->tgt->vmid);
+
+	infop->vmid = lup->tgt->vmid;
+	infop->vmdkid = vmdkid;
+	infop->vmdk = GetVmdkHandle(vmdkid);
+	assert(infop->vmdk != kInvalidVmdkHandle);
+
+	rc = pthread_mutex_init(&infop->lock, NULL);
+	assert(rc == 0);
+
+	DLL_INIT(&infop->sched_cmd_list);
+
+	/** TODO: Use a symbolic const over here */
+	infop->nr_results = 32;
+	infop->request_resultsp = calloc(infop->nr_results,
+					sizeof(*infop->request_resultsp));
+	if (!infop->request_resultsp) {
```

Comment:
- if calloc fails you have memory and resource leak

- free(vmdkid)
- pthread_mutex_Destroy()

Reply:
done.